Add kubelet plugin manager

taragu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**: 
Implement a controller that manages plugin registration/unregistration

**Which issue(s) this PR fixes**:
Fixes #73371

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Kubelet plugin registration now has retry and exponential backoff logic for when registration of plugins (like CSI or device plugin) fail.
```

taragu commented:
```
* "reconciler" should use the [`nestedpendingoperations` library](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go) to execute registration and unregistration operations. This library automatically takes care of exponential backoff and ensures no more then one operation (registration or unregistration) can happen at a given time for a given socket.
```
@saad-ali I was looking at nestedpendingoperations but wasn't sure how to use it to register/unregister plugin. Could you please point me to a similar example?

taragu commented:
cc @vikaschoudhary16

saad-ali commented:
@taragu there is a generic version of the `nestedpendingoperation` that is simpler and not volume specific called `goroutinemap`. See https://github.com/kubernetes/kubernetes/blob/master/pkg/util/goroutinemap/goroutinemap.go (edited) 

The `operation_executor` is just a wrapper around `nestedpendingoperation` (or `goroutinemap`). I would define a new `operation_executor` for the plugin manager that defines two operations: `RegisterPlugin()` and `UnregisterPlugin()`. This `operation_executor` would use `goroutinemap` to ensure that no more then one operation starts for the same `socket`. So the key that you use in the `goroutinemap` (I think) would be the socket path.

Once you have an `operation_executor` defined, you can create two caches: `desired` and `actual` that keep track of the sockets that we want to register and the sockets that we actually have successfully registered.

Then you can create the dumb diff loop, the `reconciler`, it will (maybe every 100 ms) diff `desired` and `actual`. If a socket is in `desired` but not in `actual` it will call `operation_executor.Register()`. If a socket is in `actual` but not `desired` it will call `operation_executor.Unregister()`. And because `operation_executor` uses `goroutinemap` it will refuse to start a new register or unregister operation while one is pending (and automatically handle exponential backoff).

Once you have that, you need to write a `populator` -- in this case that will basically just be the existing code that currently calls the old Register method -- and instead it would add to the desired state cache on register, and remove from desired state cache on unregister (populator in this case doesn't need to be a new package).

taragu commented:
@saad-ali thank you so much!! This is very comprehensive. I'll try to digest this later this evening.

taragu committed:
Implement plugin manager - a controller that manages plugin registration/unregistration

saad-ali commented on pkg/kubelet/pluginmanager/cache/desired_state_of_world.go:
```diff
@@ -0,0 +1,206 @@
+		if reflect.DeepEqual(foundInfo, pluginInfo) {
+			return nil
+		}
+		klog.V(2).Infof("Plugin (Name %s, Type %s, Path %s) exists in desired state cache, but the plugin information is different; updating plugin info", pluginInfo.Name, pluginInfo.Type, pluginInfo.SocketPath)
```
So I guess the idea here is that if a plugin get's upgraded. The socket remains the same. But the information (like "supported version" changes. When that happens, since the object in desired state is different from actual state, reconciler will trigger a Unregister operation for that socket, and once the unregister completes, it will start the register operation with the new plugin info (because reconciler starts unregister before register and never lets more then one operation happen at once for the same socket).

That sounds like the right way to handle it. But I'll think over this some more.

CC @msau42 for FYI

saad-ali replied:
Ok this logic makes sense to me. It's very nice way of handling the case where a plugin mutates (changes the versions it supports, for example).

Only follow up here:

1) Capture this in a comment here, so it is clear that is the intention.
2) If it is not super painful, add a test that will do this:
    * Create a socket which returns supported versions `0.3.0`.
    * Verify that Register is called correctly to register it.
    * Replace the socket with a socket that returns supported versions `0.3.0, 1.0.0`
    * Verify that the socket is first unregistered and then reregistered.

You can do this in an E2E test in a follow up PR if you want, just make sure to open a tracking bug for it.

taragu replied:
@saad-ali After my recent update this will longer be true. If a plugin is upgraded while socket remains the same, the plugin will not be re-registered. I made this change because I needed to move the `GetInfo` logic into `operation_generator.go`. Also, the existing `Start(...)` logic of `plugin_watcher.go` trigger registration/deregistration on file create and delete event, and I'm leaving that logic untouched.

saad-ali replied:
We were able to reintroduce this logic via timestamps.

saad-ali commented on pkg/kubelet/util/pluginwatcher/BUILD:
```diff
@@ -6,17 +6,16 @@ go_library(
         "example_handler.go",
         "example_plugin.go",
         "plugin_watcher.go",
```
Let's go ahead and move this entire package from `pkg/kubelet/util/pluginwatcher/*` to `pkg/kubelet/pluginmanager/pluginwatcher/*` so that all the plugin code is in one place.

saad-ali commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -26,69 +26,40 @@ import (
 // deprecatedSockDir refers to a pre-GA directory that was used by older plugins
 // for socket registration. New plugins should not use this directory.
-func NewWatcher(sockDir string, deprecatedSockDir string) *Watcher {
+func NewWatcher(sockDir string, deprecatedSockDir string, desiredStateOfWorld cache.DesiredStateOfWorld) *Watcher {
```
nit: It would be really nice if `NewWatcher(...)` returned a Watcher *interface* instead of a struct (that would make it easier to test). If you have time, and want to, you can make this change.

saad-ali replied:
>  I plan to address with followup PRs:
>    * having NewWatcher(...) return a Watcher interface

SGTM. Could you open up a P2 cleanup issue to track this work?

taragu replied:
Sure. Just created #76539

saad-ali commented on pkg/kubelet/kubelet.go:
```diff
@@ -786,9 +788,16 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
-			klet.getPluginsRegistrationDir(), /* sockDir */
-			klet.getPluginsDir(),             /* deprecatedSockDir */
+		desiredStateOfWorld := plugincache.NewDesiredStateOfWorld()
+		klet.pluginManager = pluginmanager.NewPluginManager(
```
1. Let's move the construction of `NewWatcher()`, `NewDesiredStateOfWorld()`, and `NewActualStateOfWorld()` inside the `NewPluginManager(...)` constructor method instead of passed in here as parameters -- these are implementation details of the PluginManager that are not relevant to the kubelet.
2. Just have the kubelet pass in `sockDir`, `deprecatedSockDir`. And maybe `recorder` (Double check if PluginManger should be constructing its own new recorder or using this one. I don't remember off the top of my head).

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,160 @@
+	return plugintypes.GeneratedOperations{
+		OperationFunc:     registerPluginFunc,
+		CompleteFunc:      nil, // TODOTARA NOT SURE IF THIS IS NEEDED YET
+		EventRecorderFunc: nil, // TODOTARA NOT SURE IF THIS IS NEEDED YET
```
I need to double check this. Go ahead and address the existing feedback, and let me get back to you on this.

saad-ali replied:
For volumes, the `EventRecorderFunc` is a nice way to ensure that for operation like attach/detach/mount/unmount, a Kubernetes event is generated on success or failure. Instead of having each operation sprinkle the event generation code everywhere, you let the lower level operation executor handle it in a common way.

Adding Kubernetes events to the node object for failed (and successful) plugin registration would be super cool, but you don't have to do it in the PR since it doesn't exist yet. Let's open up an issue to track that as separate follow up work.

Similarly, for volumes, the `CompleteFunc` is a nice way to ensure that metrics are emitted for the attach/detach/mount/unmount operations in a standard way instead of having that code sprinkled inside every operation. 

It looks like you added a metrics package in this PR, but none of the operations are wired up to emit any events yet. Adding metrics for plugin operations (Register/Unregister) would be super cool, but given the functionality doesn't exist yet, it's up to you if you want to add it in this PR. I would be ok with you opening an issue to track that as separate, followup work.

saad-ali replied:
> I plan to address with followup PRs:
>   * wiring metrics for plugin operations

SGTM. Can you open up an issue to track this work.

taragu replied:
Sure, I've created #76538

taragu commented:
@saad-ali Thanks very much for your review comments! I've addressed all of your comments except the following two, which I plan to address with followup PRs:
- having NewWatcher(...) return a Watcher interface
- wiring metrics for plugin operations

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,150 @@
+	GenerateRegisterPluginFunc(
+		pluginToRegister cache.PluginInfo,
+		pluginHandler cache.PluginHandler,
+		actualStateOfWorldUpdater ActualStateOfWorldUpdater) (func() error, error)
```
Drop the `error` from the return parameters of the `Generate...` methods. When there is an option to return an error, people will start adding validation to the `Generate...` methods. And the problem with that is the reconciler is dumb, it will just call the same method again, and that will fail, and that will cause the logs to be filled with the same error again and again.

Instead we should encourage people to do all the work, including the validation inside the function that is returned, therefore if validation fails, the `goroutinemap` framework handles exponential back off. (See https://github.com/kubernetes/kubernetes/pull/71581/files for an example of this change.)

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go:
```diff
@@ -0,0 +1,120 @@
+
+// Package operationexecutor implements interfaces that enable execution of
+// register and unregister operations with a
+// nestedpendingoperations so that more than one operation is never triggered
```
nit: `nestedpendingoperations` -> `goroutinemap`

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go:
```diff
@@ -0,0 +1,120 @@
+// Package operationexecutor implements interfaces that enable execution of
+// register and unregister operations with a
+// nestedpendingoperations so that more than one operation is never triggered
+// on the same plugin for the same pod.
```
drop `for the same pod`.

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go:
```diff
@@ -0,0 +1,120 @@
+)
+
+// OperationExecutor defines a set of operations for registering and unregistering
+// a plugin that are executed with a NewNestedPendingOperations which
```
nit: `NewNestedPendingOperations` -> `NewGoRoutineMap`

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/nestedpendingoperations/nestedpendingoperations.go:
```diff
@@ -0,0 +1,271 @@
+execute in parallel for plugins on the same pod as long as they are operating on
+different sockets.
+*/
+package nestedpendingoperations
```
Resolved.

saad-ali commented on pkg/kubelet/pluginmanager/plugin_manager.go:
```diff
@@ -0,0 +1,132 @@
+
+	// GetRegisteredPlugins generates and returns a list of plugins that are
+	// successfully registered plugins in the current actual state of world.
+	GetRegisteredPlugins() []cache.PluginInfo
```
This method is only used by `plugin_manager_test.go`. That indicates the unit test is making assumptions about implementation it should not.

To fix this, let's do this:
1. Remove `GetRegisteredPlugins()` from this interface.
2. Create a "mock" or "fake" `pluginHandler cache.PluginHandler` that keeps track of when (and maybe how many times) "Validate", "Register" or "Unregister" are called for each socket, and the fake should expose methods that allow the tester to inspect that at some point after construction and execution. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager_fake.go for a rough example of this.
3. Modify `plugin_manager_test.go` so that instead of using `GetRegisteredPlugins()` to figure out if a plugin is registered, it uses the "fake" or "mock" you created above.

I know this is a non-trivial amount of work, and if you do not want to do it now, make sure to open an issue to track the work so it gets addressed in a follow up.

saad-ali commented on pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+type actualStateOfWorld struct {
```
Add a compile type interface implementation check:

```Go
var _ ActualStateOfWorld = &actualStateOfWorld{}
```

saad-ali commented on pkg/kubelet/pluginmanager/cache/desired_state_of_world.go:
```diff
@@ -0,0 +1,180 @@
+	}
+}
+
+type desiredStateOfWorld struct {
```
Add a compile type interface implementation check:

```Go
var _ DesiredStateOfWorld = &desiredStateOfWorld{}
```

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go:
```diff
@@ -0,0 +1,120 @@
+	RemovePlugin(socketPath string)
+}
+
+type operationExecutor struct {
```
Add a compile type interface implementation check:

```Go
var _ OperationExecutor = &operationExecutor{}
```

saad-ali commented on pkg/kubelet/pluginmanager/plugin_manager.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+// pluginManager implements the PluginManager interface
+type pluginManager struct {
```
Add a compile type interface implementation check:

```Go
var _ PluginManager = &pluginManager{}
```

saad-ali commented on pkg/kubelet/pluginmanager/reconciler/reconciler.go:
```diff
@@ -0,0 +1,153 @@
+	}
+}
+
+type reconciler struct {
```
Add a compile type interface implementation check:

```Go
var _ Reconciler = &reconciler{}
```

saad-ali commented on pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go:
```diff
@@ -275,151 +237,28 @@ func (w *Watcher) handlePluginRegistration(socketPath string) error {
 		return fmt.Errorf("failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
 	}
 
```
We should move all the above code (`dial`, `context.WithTimeout`, and `GetInfo`) to the new Register operation as well. If we leave it here, and it fails, we will never retry. If it is part of the new Register operation, the reconciler+goroutinemap will handle retries with exponential backoff.

taragu replied:
@saad-ali That makes sense. That means we'll also need to update the input around `DesiredStateOfWorld`'s `AddOrUpdatePlugin` and `ActualStateOfWorld` since we won't have the full plugin info before we call `GenerateRegisterPluginFunc`

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,150 @@
+			return fmt.Errorf("pluginHandler.RegisterPluginFunc failed")
+		}
+		// // Notify is called after register to guarantee that even if notify throws an error Register will always be called after validate
+		// if err := og.notifyPlugin(client, true, ""); err != nil {
```
We should still do this. Essentially this is a call back to the plugin to say, hey the handler for your type says you are not a valid plugin. So heads up, registration failed.

saad-ali commented on pkg/kubelet/pluginmanager/pluginwatcher/example_handler.go:
```diff
@@ -26,8 +26,8 @@ import (
-	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta2"
 )
```
Update the "gRPC Service Lifecycle" section of `pluginmanager/pluginwatcher/README.md` file to indicate that if any of the steps in registration fails, on retry registration will start from scratch.
For example, imagine for a new socket:

* Registration.GetInfo is called against socket. Succeeds.
* Validate is called against internal plugin type handler. Succeeds.
* Register is called against internal plugin type handler. Fails.
* NotifyRegistrationStatus is called against socket to indicate registration failed. Succeeds.

After some back off time time, the `pluginmanager` will try again from the begining:

* GetInfo is called against socket, again. Succeeds.
* Validate is called against internal plugin type handler again. Succeeds.
* Register is called against internal plugin type handler again. Succeeds this time.
* NotifyRegistrationStatus is called against socket to indicate registration succeed. Succeeds.
* Plugin is successfully registered.

May also be a good to document what happens if for `GetInfo` call to socket returns a different "type", "endpoint" or "supported version" then a previous successful registration for the same "socket path" in the readme.

taragu replied:
@saad-ali I'm finishing off the last bit of changes but this will no longer be true:
```
what happens if for GetInfo call to socket returns a different "type", "endpoint" or 
"supported version" then a previous successful registration for the same 
"socket path"
```
Because `GetInfo` is now called in the register and unregister calls (`GenerateRegisterPluginFunc` and `GenerateUnregisterPluginFunc`), I've updated the actual/desired state caches so they no longer keep information about plugins except for the socket path and `foundInDeprecatedDir` (I haven't pushed the changes yet). As a result, we won't be able to check if a plugin exists for the same socket path but other information has changed. I think this is okay because `plugin_watcher` only trigger plugin creation when a new file is detected, so the scenario of plugin information changing for an existing socket file won't happen since I'm leaving the [`Start()` method](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/pluginwatcher/plugin_watcher.go#L89) untouched

saad-ali replied:
The problem is that if a file is deleted and recreated very quickly, the reconciler may not even notice that it was removed from the desired state. So reregistration won't happen.

@msau42 had a great suggestion for how we can still retain the cool reregisteration behavior:

1. Generate and store current timestamp during `desiredStateOfWorld.AddPlugin(...)` to the `desiredStateOfWorld` struct.
2. Upon successful registration during `actualStateOfWorld.AddPlugin(...)` copy the timestamp from `desiredStateOfWorld` to the `actualStateOfWorld` struct.
3. Have `desiredStateOfWorld.PluginExists(...)` verify that the timestamps in desired and actual state match, if not return false.

This way, after a successful registration, when the socket is deleted and recreated, it will force a reregistration because the timestamp of the socket in the desired state will be newer then the timestamp of the socket in the actual state. Specifically reconciler will first check for volume plugins to unregister, it will run `desiredStateOfWorld.PluginExists` which will return false. This will start a unregistration operation. Once the unregsitration is complete the socket will be removed from actual state. The reconciler will see the desired state has a socket that actual state doesn't and will start a registration operation.

taragu replied:
@saad-ali @msau42 how do we determine whether to add a plugin to the desired state of world cache when traversing directories in `plugin_watcher`? In the case when the socket is deleted (and the socket is not removed from the desired state cache because it happens very quickly) and recreated, if we do what you suggested in number 3

> Have `desiredStateOfWorld.PluginExists(...)` verify that the timestamps in desired and actual state match, if not return false.

when we are traversing directories in `plugin_watcher`, in `handlePluginRegistration(...)` `desiredStateOfWorld.PluginExists(...)` will return true because the timestamps in desired and actual state match, and a reregistration won't eventually happen.

saad-ali replied:
@taragu, I forgot that `pluginwatcher` called `desiredStateOfWorld.PluginExists(...)` before calling `desiredStateOfWorld.AddPlugin(...)`. What if we modified `handlePluginRegistration` in `pluginwatcher` to always call `desiredStateOfWorld.AddPlugin(...)`. And rename `desiredStateOfWorld.AddPlugin(...)` to `desiredStateOfWorld.AddOrUpdatePlugin(...)`. And then have the method update the time stamp?

taragu commented:
/test pull-kubernetes-dependencies

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,199 @@
+	registerPluginFunc := func() error {
+		client, conn, err := dial(socketPath, 10*time.Second)
+		if err != nil {
+			return fmt.Errorf("dial failed at socket %s, err: %v", socketPath, err)
```
Ditto for the rest of the errors below and similar for Unregister as well.

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,199 @@
+	actualStateOfWorldUpdater ActualStateOfWorldUpdater) func() error {
+
+	registerPluginFunc := func() error {
+		client, conn, err := dial(socketPath, 10*time.Second)
```
Ditto for the other durations in this file.

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,199 @@
+			FoundInDeprecatedDir: foundInDeprecatedDir,
+		})
+		if err := handler.RegisterPlugin(infoResp.Name, infoResp.Endpoint, infoResp.SupportedVersions); err != nil {
+			if err = og.notifyPlugin(client, false, fmt.Sprintf("plugin registration failed with err: %v", err)); err != nil {
```
Why not just `return og.notifyPlugin(...)`?

saad-ali commented on pkg/kubelet/pluginmanager/plugin_manager.go:
```diff
@@ -0,0 +1,130 @@
+	pm.reconciler.AddHandler(pluginType, handler)
+}
+
+func (pm *pluginManager) GetRegisteredPlugins() []cache.PluginInfo {
```
Remember to delete this method.

taragu commented:
/test pull-kubernetes-integration

saad-ali commented on pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:
```diff
@@ -0,0 +1,127 @@
+type PluginInfo struct {
+	SocketPath           string
+	FoundInDeprecatedDir bool
+	Timestamp            time.Time
```
This is super nit picky, but this code is so well factored, this might be worth fixing (it will discourage others from misusing it in the future).

`Timestamp` is an internal implementation detail of actual and desired state. But because we require external callers to pass in the `PluginInfo` object they are forced to set it. Instead, it would be much cleaner if we did the following:

1. Change `desiredStateOfWorld.AddOrUpdatePlugin(pluginInfo PluginInfo)` to only take the required parameters: `desiredStateOfWorld.AddOrUpdatePlugin(socketPath string, foundInDeprecatedDir bool)` (no `timestamp`).
2. Change the implementation of `desiredStateOfWorld.AddOrUpdatePlugin(socketPath string, foundInDeprecatedDir bool)` to automatically set `plugininfo.Timestamp` to `time.Now()` for `dsw.socketFileToInfo[socketPath]`.
3. Update the callers for `desiredStateOfWorld.AddOrUpdatePlugin(...)` so they no longer have to pass the full `PluginInfo` object (or the `Timestamp`).

Also, if you're feeling really inspired and want to get really fancy with separation of concerns, we could also rename `PluginInfo` to `pluginInfo` to make it [private](https://golang.org/ref/spec#Exported_identifiers) to the `cache` package. But that's a much bigger change, because it is being used all over the place.

All of this is totally optional, if you want to, if not, no worries.

saad-ali commented on pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,268 @@
+					klog.Errorf("fsWatcher received error: %v", err)
+				}
+				continue
+			case <-stopCh:
```
Ah this was in the `Stop()` method before to give all pending go routines a chance to complete.

saad-ali commented on pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,267 @@
+
+	// Traverse plugin dir after starting the plugin processing goroutine
+	if err := w.traversePluginDir(w.path); err != nil {
+		klog.Errorf("failed to traverse plugin socket path %q, err: %v", w.path, err)
```
Previously for both `w.traversePluginDir(w.path)` and `w.traversePluginDir(w.deprecatedPath)` an error would be terminal:

https://github.com/kubernetes/kubernetes/blob/28172ec17c088527bf543e65d2a2019ba7add20a/pkg/kubelet/util/pluginwatcher/plugin_watcher.go#L142-L154

Looks like you changed it to log and continue? I want to make sure I understand the reason behind the change.

taragu replied:
@saad-ali thanks for catching this. I'm actually a little unsure about my change. Currently, kubelet is in charge of starting the plugin watcher, and not in charge of stopping it. In this PR, plugin manager will now in charge of starting and stopping the watcher. So now, the kubelet will start the plugin manager, and plugin manager will start/stop the reconciler and the plugin watcher. 

I think this is better than stopping the watcher if an error is encountered when traversing the directories because I don't see a way for the kubelet to restart the watcher again. Does it make sense for the watcher to stop if the rest of kubelet is still running? 

Also this approach is consistent with the current volume manager. Please let me know what you think.

saad-ali replied:
> I think this is better than stopping the watcher if an error is encountered when traversing the directories because I don't see a way for the kubelet to restart the watcher again. Does it make sense for the watcher to stop if the rest of kubelet is still running?

That seems ok to me. Log an error and continue instead of completely giving up.

Looks like https://github.com/kubernetes/kubernetes/pull/75110 recently merged. Can you make sure this PR doesn't undo those changes. I don't know how relevant that bug is with this refactor, but lets minimize the unnecessary change unless we have a good reason for it.

saad-ali commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,204 @@
+	unregisterPluginFunc := func() error {
+		client, conn, err := dial(socketPath, dialTimeoutDuration)
+		if err != nil {
+			return fmt.Errorf("dial failed at socket %s, err: %v", socketPath, err)
```
nit: Prefix error string with `UnregisterPlugin error -- dial failed...` like you did for Register already.

Ditto for the other errors in this method.

saad-ali reviewed (commented):
Thank you! This is looking great.

Looks like you also need a rebase. Should be simple just the `hack/.golint_failures` file.

saad-ali commented on pkg/kubelet/pluginmanager/BUILD:
```diff
@@ -0,0 +1,56 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
```
Oh and before I forget, since this is a significant new component, add an OWNERS file at `pkg/kubelet/pluginmanager/OWNERS` and add yourself and me (`saad-ali`) as approvers. Example: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/OWNERS

saad-ali reviewed (commented):
One last comment. And feel free to squash.

taragu commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

taragu commented:
/test pull-kubernetes-integration

saad-ali commented:
/test pull-kubernetes-integration

saad-ali commented on pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,267 @@
+			if err := w.handleCreateEvent(event); err != nil {
+				klog.Errorf("error %v when handling create event: %s", err, event)
+			}
+			// w.wg.Add(1)
```
nit: remove commented out code.

saad-ali reviewed (commented):
Thanks for all the hard work on this.

One last nit and PR needs rebase.

Otherwise LGTM
/lgtm
/approve

saad-ali commented:
/assign @vishh for approval

saad-ali reviewed (commented):
@taragu, one more thing I forgot, we should undo the temporary fix from https://github.com/kubernetes/kubernetes/pull/72873 as well. We could do that in a follow up PR if you want.

taragu commented:
@saad-ali sounds good. I will undo #72873 in a followup PR

msau42 commented:
/hold
Spoke with @vishh. There are still some discussions happening with @saad-ali  around if we can just have the drivers restart

vishh commented:
/approve 
(after offline discussion with @saad-ali who patiently explained that k'let can silently fail re-registration when the plugins assume normal operation)

saad-ali commented:
/lgtm

saad-ali commented:
/milestone v1.15

saad-ali commented:
/lgtm

taragu commented:
/hold cancel

taragu commented:
/test pull-kubernetes-integration

taragu commented:
/test pull-kubernetes-integration

tedyu commented on pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:
```diff
@@ -0,0 +1,127 @@
+	asw.Lock()
+	defer asw.Unlock()
+
+	if _, ok := asw.socketFileToInfo[socketPath]; ok {
```
Not sure why the check is needed - we can delete the entry directly.
It is fine if the key doesn't exist.

tedyu commented on pkg/kubelet/pluginmanager/cache/desired_state_of_world.go:
```diff
@@ -0,0 +1,172 @@
+	dsw.Lock()
+	defer dsw.Unlock()
+
+	if _, ok := dsw.socketFileToInfo[socketPath]; ok {
```
No need to check prior to deletion

tedyu commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,204 @@
+	return nil
+}
+
+// Dial establishes the gRPC communication with the picked up plugin socket. https://godoc.org/google.golang.org/grpc#Dial
```
Should be lowercase: dial

tedyu commented on pkg/kubelet/pluginmanager/reconciler/reconciler.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+func (rc *reconciler) getHandlers() map[string]cache.PluginHandler {
+	rc.Lock()
```
Getting read lock should suffice

lujinda commented on pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go:
```diff
@@ -0,0 +1,204 @@
+	actualStateOfWorldUpdater ActualStateOfWorldUpdater) func() error {
+
+	unregisterPluginFunc := func() error {
+		client, conn, err := dial(socketPath, dialTimeoutDuration)
```
If the plugin has already exited, the dial here will fail.

tedyu replied:
The client is needed to retrieve from pluginHandlers using infoResp.Type

Do you have suggestion on how to get plugin handler in that case ?

lujinda replied:
> The client is needed to retrieve from pluginHandlers using infoResp.Type
> 
> Do you have suggestion on how to get plugin handler in that case ?

The csi plugin will delete the sock file when it exits. When the operationExecutor.UnregisterPlugin is called, the sock file is not existed. 
Can add a sock file to the plugin mapping table.   ≧◠◡◠≦  Just like 1.14:
```

func (w *Watcher) getPlugin(socketPath string) (pathInfo, bool) {
	w.mutex.Lock()
	defer w.mutex.Unlock()

	plugin, ok := w.plugins[socketPath]
	return plugin, ok
}
```

tedyu replied:
@taragu 
What do you think of the above suggestion ?

k2huang replied:
From my POV, we can save socketPath: infoResp.Type in actualStateOfWorld on RegisterPlugin phase, and then it can be used on UnregisterPlugin phase

tedyu replied:
I made similar suggestion over #87282

Let's see what other people think.
